### aim: put medusa files in an appropriate format for further look - one file with appropriate variables per year

### date: Dec 12, 2022

In [1]:
from IPython.display import HTML

HTML('''<script>
code_show=true; 
function code_toggle() {
 if (code_show){
 $('div.input').hide();
 } else {
 $('div.input').show();
 }
 code_show = !code_show
} 
$( document ).ready(code_toggle);
</script>
<form action="javascript:code_toggle()"><input type="submit" value="Click here to toggle on/off the raw code."></form>''')


In [2]:
%reset

Once deleted, variables cannot be recovered. Proceed (y/[n])?  y


In [3]:
import numpy as np
from cmocean import cm
import cartopy as cp
import cartopy.crs as ccrs
import netCDF4 as nc
import matplotlib.pyplot as plt
import xarray as xr
import sys
sys.path.append('/gpfs/home/mep22dku/scratch/SOZONE')
#list of models
sys.path.append('/gpfs/home/mep22dku/scratch/SOZONE/UTILS')
import lom
import utils as ut

%matplotlib inline
import warnings
from datetime import datetime
warnings.filterwarnings('ignore')
import cartopy.feature as cfeature
from importlib import reload
import matplotlib.path as mpath
import glob
import pickle
import pandas as pd
import seawater
import time
plt.rcParams.update({'font.size': 12})
font = {'family' : 'normal',
'weight' : 'normal',
'size'   : 12}

plt.rc('font', **font)

In [4]:
tlist = glob.glob('/gpfs/data/greenocean/software/resources/MEDUSA/oneyear2009ukesm/*20090401-*.nc')

#print(tlist)
for t in tlist:
    ds = xr.open_dataset(t)
    print(t)
    tvars = list(ds.keys())
    print(tvars)
    print('--')
    


/gpfs/data/greenocean/software/resources/MEDUSA/oneyear2009ukesm/medusa_bc370o_1m_20090401-20090501_ptrd-T.nc
['bounds_lon', 'bounds_lat', 'area', 'deptht_bounds', 'thkcello', 'time_centered_bounds', 'SMS_DIN_E3T', 'SMS_SIL_E3T', 'SMS_FER_E3T', 'SMS_DiC_E3T', 'SMS_ALK_E3T', 'TOT_DIN_E3T', 'TOT_SIL_E3T', 'TOT_FER_E3T', 'TOT_DiC_E3T', 'TOT_ALK_E3T']
--
/gpfs/data/greenocean/software/resources/MEDUSA/oneyear2009ukesm/medusa_bc370o_1m_20090401-20090501_diad-T.nc
['bounds_lon', 'bounds_lat', 'area', 'deptht_bounds', 'thkcello', 'time_centered_bounds', 'INVTN', 'INVTSI', 'INVTFE', 'PRN', 'MPN', 'PRD', 'MPD', 'OPAL', 'OPALDISS', 'GMIPn', 'GMID', 'MZMI', 'GMEPN', 'GMEPD', 'GMEZMI', 'GMED', 'MZME', 'DETN', 'MDET', 'AEOLIAN', 'BENTHIC', 'SCAVENGE', 'PN_JLIM', 'PN_NLIM', 'PN_FELIM', 'PD_JLIM', 'PD_NLIM', 'PD_FELIM', 'PD_SILIM', 'PDSILIM2', 'SDT__100', 'SDT__200', 'SDT__500', 'SDT_1000', 'TOTREG_N', 'TOTRG_SI', 'REG__100', 'REG__200', 'REG__500', 'REG_1000', 'FASTN', 'FASTSI', 'FASTFE', 'FASTC', '

In [5]:
dtype = 'ptrd-T'
t = f'/gpfs/data/greenocean/software/resources/MEDUSA/oneyear2009ukesm/medusa_bc370o_1m_20090401-20090501_diad-T.nc'

ds = xr.open_dataset(t)
print(ds['ML_PRN'])

<xarray.DataArray 'ML_PRN' (time_counter: 1, y: 332, x: 362)>
[120184 values with dtype=float32]
Coordinates:
    nav_lat        (y, x) float32 ...
    nav_lon        (y, x) float32 ...
    time_centered  (time_counter) object ...
  * time_counter   (time_counter) float64 0.0
Dimensions without coordinates: y, x
Attributes:
    long_name:           Non-diatom mixed layer PP
    units:               mmol-N/m2/d
    online_operation:    average
    interval_operation:  2700 s
    interval_write:      1 month
    cell_methods:        time: mean (interval: 2700 s)
    cell_measures:       area: area


In [10]:
ds['EXPC3']

<xarray.DataArray 'EXPC3' (time_counter: 1, deptht: 75, y: 332, x: 362)>
[9013800 values with dtype=float32]
Coordinates:
    nav_lat        (y, x) float32 ...
    nav_lon        (y, x) float32 ...
  * deptht         (deptht) float32 0.50576 1.5558553 ... 5698.0605 5902.0576
    time_centered  (time_counter) object ...
  * time_counter   (time_counter) float64 0.0
Dimensions without coordinates: y, x
Attributes:
    long_name:           Downward Flux of Particulate Org Carbon
    units:               mmol-C/m3/d
    online_operation:    average
    interval_operation:  2700 s
    interval_write:      1 month
    cell_methods:        time: mean (interval: 2700 s)
    cell_measures:       area: area

In [9]:
dtype = 'ptrd-T'
t = f'/gpfs/data/greenocean/software/resources/MEDUSA/oneyear2009ukesm/medusa_bc370o_1m_20090401-20090501_diad-T.nc'

ds = xr.open_dataset(t)
print(t)
tvars = list(ds.keys())

for tv in tvars:
    print(tv)
    try:

        print(ds[tv].long_name)
        print(ds[tv].units)
        print()
    except:
        print()

/gpfs/data/greenocean/software/resources/MEDUSA/oneyear2009ukesm/medusa_bc370o_1m_20090401-20090501_diad-T.nc
bounds_lon

bounds_lat

area

deptht_bounds

thkcello
Ocean Model cell Thickness
m

time_centered_bounds

INVTN
Column N inventory
mmol-N/m2

INVTSI
Column Si inventory
mmol-Si/m2

INVTFE
Column Fe inventory
mmol-Fe/m2

PRN
Non-diatom primary production
mmol-N/m2/d

MPN
Non-diatom non-grazing losses
mmol-N/m2/d

PRD
Diatom primary production
mmol-N/m2/d

MPD
Diatom non-grazing losses
mmol-N/m2/d

OPAL
Diatom biogenic opal production
mmol-Si/m2/d

OPALDISS
Diatom biogenic opal dissolution
mmol-Si/m2/d

GMIPn
Microzoo grazing on non-diatoms
mmol-N/m2/d

GMID
Microzoo grazing on detritus
mmol-N/m2/d

MZMI
Microzoo non-grazing losses
mmol-N/m2/d

GMEPN
Mesozoo grazing on non-diatoms
mmol-N/m2/d

GMEPD
Mesozoo grazing on non-diatoms
mmol-N/m2/d

GMEZMI
Mesozoo grazing on microzoo
mmol-N/m2/d

GMED
Mesozoo grazing on detritus
mmol-N/m2/d

MZME
Mesozoo non-grazing losses
mmol-N/m2/d



In [ ]:
td = xr.open_dataset('/gpfs/data/greenocean/software/resources/MEDUSA/oneyear2009ukesm/nemo_bc370o_1m_20090401-20090501_grid-V.nc')
#print(td)

nav_lat = td['nav_lat'][:,:].values
nav_lon = td['nav_lon'][:,:].values
depthv = td['depthv'].values
time_centered = td['time_centered'].values
time_counter = td['time_counter'].values

# tvars = list(td.keys())
# # w['CO2FLUX']
# print(tvars)

# td['vo']

vomecrty = td['vo']


savenam = 'test2.nc'
data_vars = {'vomecrty':(['time_counter', 'depthv', 'y', 'x'], vomecrty.values,
{'units': 'µatm',
 'long_name':'Surface ocean pCO2'}),
}
# define coordinates
coords = {'time_counter': (['time_counter'], time_counter),
          'time_centered': (['time_centered'], time_centered),
'nav_lat': (['y','x'], nav_lat),
'nav_lon': (['y','x'], nav_lon),
'depthv': (['depthv'], depthv)}


# define global attributes
attrs = {'made in':'SOZONE/MEDUSA/makeYearlyMEDUSAsubsetfiles.ipynb',
'desc': 'yearly medusa files, saving only variables of interest'
}
ds = xr.Dataset(data_vars=data_vars,
coords=coords,
attrs=attrs)

ds.to_netcdf(savenam)

In [ ]:
td2 = xr.open_dataset('test2.nc')

In [ ]:
mm_tom = xr.open_dataset('/gpfs/home/mep22dku/cdftools/mesh_mask3_6.nc')
mm_med = xr.open_dataset('/gpfs/home/mep22dku/cdftools/mesh_mask_eORCA1_wrk.nc')

tomvars = list(mm_tom.keys())
medvars = list(mm_med.keys())

print(np.shape(mm_med['gdepw'][:]))
print(np.shape(mm_tom['gdepw_1d'][:]))

print('meshmask variables TOM')
print(tomvars)

# print(np.shape(mm_tom['gdepw']))
# print(np.shape(mm_med['gdepw']))
''
print('meshmask variables UKESM')
print(medvars)
print()
main_list = list(set(tomvars) - set(medvars))
print('in TOM but not in UKESM')
print(main_list)
print()
main_list = list(set(medvars) - set(tomvars))
print('in UKESM but not in TOM')
print(main_list)

In [ ]:
def make_yearly_subset_nc(yr,runname):

    #what is new file going to be called:
    savenam = f'/gpfs/data/greenocean/software/resources/MEDUSA/PROCESSED/medusa_{runname}_1y_{yr}_diad-T-subset.nc'
    print(savenam)
    
    #get by-month dimension for this year:
    times = pd.date_range(f"{yr}/01/01",f"{yr+1}/01/01",freq='M',closed='left')
    #get the spatial dimensions from a variable
    q = xr.open_dataset('/gpfs/data/greenocean/software/resources/MEDUSA/diad/cj198_1950/medusa_cj198o_1m_19500101-19500201_diad-T.nc')
    nav_lat = q['nav_lat'].values
    nav_lon = q['nav_lon'].values
    deptht = q['deptht'].values
    
    ## define variables that we are saving
    OCN_PCO2 = np.zeros([12,332,362])
    CO2FLUX = np.zeros([12,332,362])
    # area = np.zeros([332,362])

    #get the files we are converting
    td = f'/gpfs/data/greenocean/software/resources/MEDUSA/diad/*/medusa_{runname}*_1m_{yr}*01-*diad-T.nc'
    fils = glob.glob(td)
    fils.sort()
    if len(fils) != 12:
        print(f'missing files for year {yr} in {runname}: we have {len(fils)}')
        return

    else:
        for i in range(0,12):
            tfil = xr.open_dataset(fils[i])
            # if i == 0:
            #     area[:,:] = tfil['area'][:,:].values
            OCN_PCO2[i,:,:] = tfil['OCN_PCO2'][:,:].values
            CO2FLUX[i,:,:] = tfil['CO2FLUX'][:,:].values



        # define data with variable attributes
        data_vars = {'OCN_PCO2':(['time_counter', 'nav_lat', 'nav_lon'], OCN_PCO2,
                                 {'units': 'µatm',
                                  'long_name':'Surface ocean pCO2'}),
                     'CO2FLUX':(['time_counter', 'nav_lat', 'nav_lon'], CO2FLUX,
                                 {'units': 'mmol-C/m2/d',
                                  'long_name':'Air-sea CO2 flux'}),   
                     # 'area':(['nav_lat', 'nav_lon'], area,
                     #             {'units': 'mmol-C/m2/d',
                     #              'long_name':'Air-sea CO2 flux'})   
                    }

        # define coordinates
        coords = {'time_counter': (['time_counter'], times),\
                 'nav_lat': (['nav_lat'], nav_lat[:,0]),\
                 'nav_lon': (['nav_lon'], nav_lon[0,:]),\
                 'deptht': (['deptht'], deptht)}

        # define global attributes
        attrs = {'made in':'SOZONE/MEDUSA/makeYearlyMEDUSAsubsetfiles.ipynb',
                'desc': 'yearly medusa files, saving only variables of interest'
                }

        ds = xr.Dataset(data_vars=data_vars,
                        coords=coords,
                        attrs=attrs)

        try:
            ds.to_netcdf(savenam)
        except:
            print(f'seems like {savenam} exists already')
    
#

## Code to list keys, to know how big file is, to know longnames

In [ ]:
import os

tw = '/gpfs/home/mep22dku/scratch/MEDUSA/*/medusa_cj198o_1m_19900101-19900201_diad-T.nc'
t = glob.glob(tw)
print(t)
q = os.stat(t[0])
print(q.st_size* 0.000001) #MB
w = nc.Dataset(t[0], "a")
#print(w)
w = xr.open_dataset(t[0])
tvars = list(w.keys())
# w['CO2FLUX']
print(tvars)

printlong = False
if printlong:
    
    for i in range(0,len(tvars)):
        print([tvars[i]])

        try:
            print(w[tvars[i]].long_name)
        except:
            print('*** no longname')

        print()


In [ ]:
# my_string="hello python world , i'm a beginner"
# print(my_string.split("world",1)[1])

In [ ]:
# cond =  ((v_name == 'bounds_lon') | \
#  (v_name == 'bounds_lat') | \
#  (v_name == 'area') | \
#  (v_name == 'deptht_bounds') | \
#  (v_name == 'thkcello') | \

#  (v_name == 'WIND') | \
#  (v_name == 'ATM_PCO2') | \
#  (v_name == 'OCN_PCO2') | \
#  (v_name == 'CO2FLUX') | \
#  (v_name == 'TCO2') | \
#  (v_name == 'TALK') | \
#  (v_name == 'C_CONS') | \
#  (v_name == 'CHL_MLD') | \
#  (v_name == 'TPP3'))


In [ ]:
q

## hardwon way to copy over variables

In [ ]:
# bdir = '/gpfs/home/mep22dku/scratch/MEDUSA/'
# tardir = 'cj198_1990/'
# tstr = f'{bdir}{tardir}/*diad-T.nc'
# print(tardir)
# q = glob.glob(tstr)
# print(len(q))
# for i in range(0,len(q)):
#     print(i)
#     dsin = nc.Dataset(q[i])
#     justfil = (q[i].split(tardir,1)[1])
#     newfil = justfil.replace("diad-T", "diad-T_SUBSET")
#     newfilhome = f'/gpfs/home/mep22dku/scratch/MEDUSA/SUBSET/{newfil}'
#     dsout = nc.Dataset(newfilhome, "w", format="NETCDF4")
#     #Copy dimensions
#     for dname, the_dim in dsin.dimensions.items():
#         dsout.createDimension(dname, len(the_dim) if not the_dim.isunlimited() else None)

#     # Copy variables
#     for v_name, varin in dsin.variables.items():
#         # print(v_name)
#         cond =  ((v_name == 'bounds_lon') | \
#              (v_name == 'bounds_lat') | \
#              (v_name == 'nav_lon') | \
#              (v_name == 'nav_lat') | \
#              (v_name == 'area') | \
#              (v_name == 'deptht_bounds') | \
#              (v_name == 'thkcello') | \

#              (v_name == 'WIND') | \
#              (v_name == 'ATM_PCO2') | \
#              (v_name == 'OCN_PCO2') | \
#              (v_name == 'CO2FLUX') | \
#              (v_name == 'TCO2') | \
#              (v_name == 'TALK') | \
#              (v_name == 'C_CONS') | \
#              (v_name == 'CHL_MLD') | \
#              (v_name == 'TPP3'))
#         #print(cond)
#         # cond = (v_name == 'WIND')
    
#         if cond:
#             outVar = dsout.createVariable(v_name, varin.datatype, varin.dimensions)
#             outVar.setncatts({k: varin.getncattr(k) for k in varin.ncattrs()})
#             outVar[:] = varin[:]
#             # print('--')
#             # print(f'**{v_name}**')
#             # print('--')
#         #print()
#     dsout.close()

# print('done')

In [ ]:
bdir = '/gpfs/data/greenocean/software/resources/medusa/'
#'cj198_1950','cj198_1960','cj198_1970',
#done: 'cj198_2000' 'cj198_2010','cj200_1990/','cj200_2000',
tardirs = [\
          'cj200_2010']


for t in range(0,len(tardirs)):
    tardir = tardirs[t]
    print(tardir)
    tstr = f'{bdir}{tardir}/*diad-T.nc'
    q = glob.glob(tstr)
    print(len(q))
    for i in range(0,len(q)):
        print(i)
        dsin = nc.Dataset(q[i])
        justfil = (q[i].split(tardir,1)[1])
        newfil = justfil.replace("diad-T", "diad-T_SUBSET")
        newfilhome = f'/gpfs/home/mep22dku/scratch/MEDUSA/SUBSET/{newfil}'
        #print(newfilhome)
        dsout = nc.Dataset(newfilhome, "w", format="NETCDF4")
        #Copy dimensions
        for dname, the_dim in dsin.dimensions.items():
            dsout.createDimension(dname, len(the_dim) if not the_dim.isunlimited() else None)

        # Copy variables
        for v_name, varin in dsin.variables.items():
            # print(v_name)
            cond =  ((v_name == 'bounds_lon') | \
                 (v_name == 'bounds_lat') | \
                 (v_name == 'nav_lon') | \
                 (v_name == 'nav_lat') | \
                 (v_name == 'area') | \
                 (v_name == 'deptht_bounds') | \
                 (v_name == 'thkcello') | \

                 (v_name == 'WIND') | \
                 (v_name == 'ATM_PCO2') | \
                 (v_name == 'OCN_PCO2') | \
                 (v_name == 'CO2FLUX') | \
                 (v_name == 'TCO2') | \
                 (v_name == 'TALK') | \
                 (v_name == 'C_CONS') | \
                 (v_name == 'CHL_MLD') | \
                 (v_name == 'TPP3'))
            #print(cond)
            # cond = (v_name == 'WIND')

            if cond:
                outVar = dsout.createVariable(v_name, varin.datatype, varin.dimensions)
                outVar.setncatts({k: varin.getncattr(k) for k in varin.ncattrs()})
                outVar[:] = varin[:]
                # print('--')
                # print(f'**{v_name}**')
                # print('--')
            #print()
        dsout.close()

In [ ]:
w = xr.open_dataset('/gpfs/home/mep22dku/scratch/MEDUSA/SUBSET/medusa_cj198o_1m_19990801-19990901_diad-T_SUBSET.nc')
# w = xr.open_dataset('/gpfs/home/mep22dku/scratch/MEDUSA/TEST/medusa_cj198o_1m_19941201-19950101_diad-T_SUBSET.nc')
plt.figure()
tw = w['CO2FLUX'][0,:,:]
plt.pcolormesh(w['CO2FLUX'][0,:,:])


w = xr.open_dataset('/gpfs/home/mep22dku/scratch/MEDUSA/SUBSET/medusa_cj198o_1m_19970101-19970201_diad-T_SUBSET.nc')
# w = xr.open_dataset('/gpfs/home/mep22dku/scratch/MEDUSA/TEST/medusa_cj198o_1m_19941201-19950101_diad-T_SUBSET.nc')
plt.figure()
plt.pcolormesh(w['CO2FLUX'][0,:,:])
tw2 = w['CO2FLUX'][0,:,:]

plt.figure()
plt.pcolormesh(tw2-tw)


In [ ]:
orig = '/gpfs/home/mep22dku/scratch/MEDUSA/cj198_1990/medusa_cj198o_1m_19941201-19950101_diad-T.nc'
dest = '/gpfs/home/mep22dku/scratch/MEDUSA/TEST/medusa_cj198o_1m_19941201-19950101_diad-T_SUBSET1.nc'

import netCDF4 as nc
import numpy as np
toexclude = ['qtrSF6', 'qintSF6']
# with nc.Dataset(orig) as src, nc.Dataset(dest, "w") as dst:
#     # copy attributes
#     for name in src.ncattrs():
#         dst.setncattr(name, src.getncattr(name))
#     # copy dimensions
#     for name, dimension in src.dimensions.iteritems():
#         dst.createDimension(
#             name, (len(dimension) if not dimension.isunlimited else None))
#     # copy all file data except for the excluded
#     for name, variable in src.variables.iteritems():
#         if name not in toexclude:
#             x = dst.createVariable(name, variable.datatype, variable.dimensions)
#             dst.variables[name][:] = src.variables[name][:]
            

import netCDF4 as nc
#toexclude = ['ExcludeVar1', 'ExcludeVar2']

with nc.Dataset(orig) as src, nc.Dataset(dest, "w") as dst:
    # copy global attributes all at once via dictionary
    dst.setncatts(src.__dict__)
    # copy dimensions
    for name, dimension in src.dimensions.items():
        dst.createDimension(
            name, (len(dimension) if not dimension.isunlimited() else None))
    # copy all file data except for the excluded
    for name, variable in src.variables.items():
        if name not in toexclude:
            x = dst.createVariable(name, variable.datatype, variable.dimensions)
            dst[name][:] = src[name][:]
            # copy variable attributes all at once via dictionary
            dst[name].setncatts(src[name].__dict__)

In [ ]:
## copying code https://github.com/SalishSeaCast/analysis_tereza/blob/master/notebooks/PI_CARBON_PAPER/PI_BOUND_COND/CLEAN/make_jdf_pi_bcs.ipynb

tw = '/gpfs/home/mep22dku/scratch/MEDUSA/*/medusa_cj198o_1m_19900101-19900201_diad-T.nc'
t = glob.glob(tw)
print(t)
q = os.stat(t[0])
print(q.st_size* 0.000001) #MB
dsin = nc.Dataset(t[0], "a")#



print(t[0])
q = t[0].replace("diad-T", "diad-T-subset8")
print(q)
dsout = nc.Dataset(q, "w", format="NETCDF4")
#Copy dimensions
for dname, the_dim in dsin.dimensions.items():
    dsout.createDimension(dname, len(the_dim) if not the_dim.isunlimited() else None)

# Copy variables
for v_name, varin in dsin.variables.items():
    #print(v_name)
    if cond:
        outVar = dsout.createVariable(v_name, varin.datatype, varin.dimensions)
        outVar.setncatts({k: varin.getncattr(k) for k in varin.ncattrs()})
        outVar[:] = varin[:]
# 
dsout.close()

In [ ]:
w = xr.open_dataset('/gpfs/home/mep22dku/scratch/MEDUSA/cj198_1990/medusa_cj198o_1m_19900101-19900201_diad-T.nc')
print(w)

In [ ]:
import h5py
os.environ['HDF5_USE_FILE_LOCKING']='FALSE'
tin = '/gpfs/home/mep22dku/scratch/MEDUSA/cj198_1990/medusa_cj198o_1m_19900101-19900201_diad-T.nc'
tout = '/gpfs/home/mep22dku/scratch/MEDUSA/cj198_1990/medusa_cj198o_1m_19900101-19900201_diad-T-subset7.nc'


#Here's what I just used and worked. @arne's answer updated for Python 3 and also to include copying variable attributes:

# import netCDF4 as nc
toexclude = ['OCN_PH', 'OCN_PCO2']

with nc.Dataset(tin) as src, nc.Dataset(tout, "w") as dst:
    # copy global attributes all at once via dictionary
    print('fuck')
    dst.setncatts(src.__dict__)
    # copy dimensions
    for name, dimension in src.dimensions.items():
        dst.createDimension(
            name, (len(dimension) if not dimension.isunlimited() else None))
    # copy all file data except for the excluded
    for name, variable in src.variables.items():
        if name not in toexclude:
            x = dst.createVariable(name, variable.datatype, variable.dimensions)
            dst[name][:] = src[name][:]
            # copy variable attributes all at once via dictionary
            dst[name].setncatts(src[name].__dict__)

In [ ]:
tlist = []
for i in range(1950,2015):
    for m in range(1,13):
        if m < 10:
            tm = f'0{m}'
        if m >= 10:
            tm = f'{m}'
        w = glob.glob(f'/gpfs/home/mep22dku/scratch/MEDUSA/*/medusa_cj198o_1m_{i}*{tm}01-*diad-T.nc')
        tlist.append((w[0]))
        
    

In [ ]:
from datetime import datetime, timedelta
w = []
w2 = []
for i in range(0, len(tlist)):
    tw = xr.open_dataset(tlist[i])

    print(len(tlist))
    q = ((tw['CO2FLUX'].time_centered.values))
    print(q[0])
    td = datetime.strptime(str(q[0]),'%Y-%m-%d %H:%M:%S')

    w.append(td)
    w2.append(2)

In [ ]:
tw = '/gpfs/home/mep22dku/scratch/MEDUSA/test_cj198/medusa_cj198o_1m_19500101-19500201_diad-T.nc'
w = xr.open_dataset(tw)
w['CO2FLUX']

In [ ]:
#             cond =  ((v_name == 'bounds_lon') | \
#                  (v_name == 'bounds_lat') | \
#                  (v_name == 'nav_lon') | \
#                  (v_name == 'nav_lat') | \
#                  (v_name == 'area') | \
#                  (v_name == 'deptht_bounds') | \
#                  (v_name == 'thkcello') | \

#                  (v_name == 'WIND') | \
#                  (v_name == 'ATM_PCO2') | \
#                  (v_name == 'OCN_PCO2') | \
#                  (v_name == 'CO2FLUX') | \
#                  (v_name == 'TCO2') | \
#                  (v_name == 'TALK') | \
#                  (v_name == 'C_CONS') | \
#                  (v_name == 'CHL_MLD') | \
#                  (v_name == 'TPP3'))

In [ ]:
rname = 'cj198'
yr = 1950
td = f'/gpfs/data/greenocean/software/resources/MEDUSA/{rname}*/*_1m_*{yr}*diad-T.nc'

w = glob.glob(td)
w.sort()
for i in range(0,len(w)):
    print(w[i])
    


<xarray.Dataset>
Dimensions:               (axis_nbounds: 2, deptht: 75, nvertex: 4, time_counter: 1, x: 362, y: 332)
Coordinates:
    nav_lat               (y, x) float32 ...
    nav_lon               (y, x) float32 ...
  * deptht                (deptht) float32 0.50576 1.5558553 ... 5902.0576
    time_centered         (time_counter) object ...
  * time_counter          (time_counter) float64 0.0
Dimensions without coordinates: axis_nbounds, nvertex, x, y
Data variables:
    bounds_lon            (y, x, nvertex) float32 ...
    bounds_lat            (y, x, nvertex) float32 ...
    area                  (y, x) float32 ...
    deptht_bounds         (deptht, axis_nbounds) float32 ...
    thkcello              (time_counter, deptht, y, x) float32 ...
    time_centered_bounds  (time_counter, axis_nbounds) object ...
    INVTN                 (time_counter, y, x) float32 ...

In [ ]:
q['CO2FLUX']

In [ ]:
def make_yearly_subset_nc(yr,runname):

    #what is new file going to be called:
    savenam = f'/gpfs/data/greenocean/software/resources/MEDUSA/PROCESSED/medusa_{runname}_1y_{yr}_diad-T-subset.nc'
    print(savenam)
    
    #get by-month dimension for this year:
    times = pd.date_range(f"{yr}/01/01",f"{yr+1}/01/01",freq='M',closed='left')
    #get the spatial dimensions from a variable
    q = xr.open_dataset('/gpfs/data/greenocean/software/resources/MEDUSA/cj198_1950/medusa_cj198o_1m_19500101-19500201_diad-T.nc')
    nav_lat = q['nav_lat'].values
    nav_lon = q['nav_lon'].values
    deptht = q['deptht'].values
    
    ## define variables that we are saving
    OCN_PCO2 = np.zeros([12,332,362])
    CO2FLUX = np.zeros([12,332,362])

    #get the files we are converting
    td = f'/gpfs/data/greenocean/software/resources/MEDUSA/{runname}*/*_1m_*{yr}*-*diad-T.nc'
    fils = glob.glob(td)
    fils.sort()
    if len(fils) != 12:
        print(f'missing files for year {yr} in {runname}: we have {len(fils)}')
        break
        
        
        
        
    else:
        for i in range(0,12):
            
    


#         # define data with variable attributes
#         data_vars = {'OCN_PCO2':(['time_counter', 'nav_lat', 'nav_lon'], OCN_PCO2,
#                                  {'units': 'µatm',
#                                   'long_name':'Surface ocean pCO2'}),
#                      'CO2FLUX':(['time_counter', 'nav_lat', 'nav_lon'], CO2FLUX,
#                                  {'units': 'mmol-C/m2/d',
#                                   'long_name':'Air-sea CO2 flux'})   
#                     }

#         # define coordinates
#         coords = {'time_counter': (['time_counter'], times),\
#                  'nav_lat': (['nav_lat'], nav_lat[:,0]),\
#                  'nav_lon': (['nav_lon'], nav_lon[0,:]),\
#                  'deptht': (['deptht'], deptht)}

#         # define global attributes
#         attrs = {'made in':'evalOutput/TA-DIC_GLODAPOBS.ipynb',
#                  'latitude bands':'band 0, -78.2 to -69.5, index 0 to 16, band 1, -69.5 to -59.3, index 16 to 28 band 2, -59.3 to -48.8, index 28 to 37',
#                 'seasons': '1 is austral summer, 4 is spring'
#                 }

#         ds = xr.Dataset(data_vars=data_vars,
#                         coords=coords,
#                         attrs=attrs)

#         print(ds)
    
make_yearly_subset_nc(1954,'cj198')


In [ ]:
print(q['TCO2'])


In [ ]:
q = xr.open_dataset('/gpfs/data/greenocean/software/resources/MEDUSA/cj198_1950/medusa_cj198o_1m_19500101-19500201_diad-T.nc')
print(q)


dic_obs = np.zeros([12,75,332,362])
# define data with variable attributes
data_vars = {'dic':(['lat_band', 'season', 'z_index'], dic_obs,
                         {'units': 'µmol/L',
                          'long_name':'free-fall velocity'}),
             'ta':(['lat_band', 'season', 'z_index'], alk_obs,
                         {'units': 'µmol/L',
                          'long_name':'free-fall velocity'})            
            }

# define coordinates
coords = {'season': (['season'], np.arange(1,5,1)),\
         'lat_band': (['lat_band'], np.arange(0,3,1)),\
         'z_index': (['z_index'], np.arange(0,31,1))}

# define global attributes
attrs = {'made in':'evalOutput/TA-DIC_GLODAPOBS.ipynb',
         'latitude bands':'band 0, -78.2 to -69.5, index 0 to 16, band 1, -69.5 to -59.3, index 16 to 28 band 2, -59.3 to -48.8, index 28 to 37',
        'seasons': '1 is austral summer, 4 is spring'
        }
            
# create dataset
ds = xr.Dataset(data_vars=data_vars,
                coords=coords,
                attrs=attrs)

fn = f'./examplenetcdf.nc'
ds.to_netcdf(fn)